x-pack/auditbeat/module/system/process/quark_provider_linux.go

```diff
@@ -0,0 +1,330 @@
+	if stats.Backend == quark.QQ_EBPF {
+		qm.log.Info("quark using EBPF")
+	} else if stats.Backend == quark.QQ_KPROBE {
+		qm.log.Info("quark using KPROBES")
+	} else {
+		qm.queue.Close()
+		qm.cachedHasher.Close()
+		return nil, fmt.Errorf("quark has an invalid backend")
+	}
+
+	return &qm, nil
+}
+
+// Run reads events from quark's queue and pushes them into output.
+// The queue is owned by this go-routine and should not be touched
+// from outside as there is no synchronization.
+func (ms *QuarkMetricSet) Run(r mb.PushReporterV2) {
+	ms.log.Info("Quark running")
+
+	metricsStamp := time.Now()
+
+MainLoop:
+	for {
+		// Poll for done
+		select {
+		case <-r.Done():
+			break MainLoop
+		default:
+		}
+
+		ms.maybeUpdateMetrics(&metricsStamp)
+
+		x := time.Now()
+		quarkEvents, err := ms.queue.GetEvents()
+		if len(quarkEvents) == 1 {
+			ms.log.Debugf("getevents took %v", time.Since(x))
+		}
+		if err != nil {
+			ms.log.Error("quark GetEvents, unrecoverable error", err)
+			break MainLoop
+		}
+		if len(quarkEvents) == 0 {
+			err = ms.queue.Block()
+			if err != nil {
+				ms.log.Error("quark Block, unrecoverable error", err)
+				break MainLoop
+			}
+			continue
+		}
+		for _, quarkEvent := range quarkEvents {
+			if !wantedEvent(quarkEvent) {
+				continue
+			}
+			if event, ok := ms.toEvent(quarkEvent); ok {
+				r.Event(event)
+			}
+		}
+	}
+
+	// Queue is owned by this goroutine, if we ever access it from
+	// outside, we need to consider synchronization.
+	ms.cachedHasher.Close()
+	ms.queue.Close()
+	ms.queue = nil
+}
+
+// toEvent converts a quark.Event to a mb.Event, returns true if we
+// were able to make an event.
+func (ms *QuarkMetricSet) toEvent(quarkEvent quark.Event) (mb.Event, bool) {
+	action, evtype := actionAndTypeOfEvent(quarkEvent)
+	process := quarkEvent.Process
+	event := mb.Event{RootFields: mapstr.M{}}
+
+	var username string
+	var processErr error
+	defer func() {
+		// Fill out root message and error.message
+		event.RootFields.Put("message",
+			makeMessage(int(process.Pid), action, process.Comm, username, processErr))
+		if processErr != nil {
+			event.RootFields.Put("error.message", processErr.Error())
+		}
+	}()
+
+	// Values that are independent of Proc.Valid
+	// Fill out event.*
+	event.RootFields.Put("event.type", evtype)
+	event.RootFields.Put("event.action", action.String())
+	event.RootFields.Put("event.category", []string{"process"})
+	event.RootFields.Put("event.kind", "event")
+	// Fill out process.*
+	event.RootFields.Put("process.name", process.Comm)
+	event.RootFields.Put("process.args", process.Cmdline)
+	event.RootFields.Put("process.args_count", len(process.Cmdline))
+	event.RootFields.Put("process.pid", process.Pid)
+	event.RootFields.Put("process.working_directory", process.Cwd)
+	event.RootFields.Put("process.executable", process.Filename)
+	if process.Exit.Valid {
+		event.RootFields.Put("process.exit_code", process.Exit.ExitCode)
+	}
+	if !process.Proc.Valid {
+		return event, true
```

Comment:
`processErr` will be `nil` here, so `makeMessage` in the deferred function will do stuff. Is that what you want? I think the answer is yes, since there is empty username handling. Just checking though.

Reply:
Correct, makeMessage does handle err being `nil`, it's basically one message if there is an error:
```
func makeMessage(pid int, action eventAction, name string, username string, err error) string {
	if err != nil {
		return fmt.Sprintf("ERROR for PID %d: %v", pid, err)
	}
...
```